datafusion-postgres/src/handlers.rs

```diff
@@ -50,72 +57,376 @@ impl PgWireServerHandlers for HandlerFactory {
+        let mut ctx = self.session_context.write().await;
+        register_current_schemas_udf(&mut ctx)?;
+        Ok(())
+    }
+
+    /// Helper function to read a custom session variable, returning the default if not set.
+    async fn session_var(&self, key: &str, default: &str) -> String {
+        self.custom_session_vars
+            .read()
             .await
+            .get(key)
+            .cloned()
+            .unwrap_or_else(|| default.to_string())
+    }
+
+    async fn handle_set(&self, variable: &ObjectName, value: &[Expr]) -> PgWireResult<()> {
+        // Join all parts of the ObjectName so that "TIME ZONE" becomes "timezone"
+        let var_name = variable
+            .0
+            .iter()
+            .map(|ident| ident.to_string())
+            .collect::<String>()
+            .to_lowercase();
+
+        let value_str = match value.first() {
+            Some(Expr::Value(v)) => match &v.value {
+                sqlparser::ast::Value::SingleQuotedString(s)
+                | sqlparser::ast::Value::DoubleQuotedString(s) => s.clone(),
+                sqlparser::ast::Value::Number(n, _) => n.to_string(),
+                _ => v.to_string(),
+            },
+            Some(expr) => expr.to_string(),
+            None => {
+                return Err(PgWireError::UserError(Box::new(
+                    pgwire::error::ErrorInfo::new(
+                        "ERROR".to_string(),
+                        "22023".to_string(),
+                        "SET requires a value".to_string(),
+                    ),
+                )));
+            }
+        };
+
+        match var_name.as_str() {
+            "timezone" => {
+                let mut sc_guard = self.session_context.write().await;
+                let mut config = sc_guard.state().config().options().clone();
+                config.execution.time_zone = Some(value_str);
+                let new_context = SessionContext::new_with_config(config.into());
+                let old_catalog_names = sc_guard.catalog_names();
+                for catalog_name in old_catalog_names {
+                    if let Some(catalog) = sc_guard.catalog(&catalog_name) {
+                        for schema_name in catalog.schema_names() {
+                            if let Some(schema) = catalog.schema(&schema_name) {
+                                for table_name in schema.table_names() {
+                                    if let Ok(Some(table)) = schema.table(&table_name).await {
+                                        new_context
+                                            .register_table(&table_name, table)
+                                            .map_err(|e| PgWireError::ApiError(Box::new(e)))?;
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+                *sc_guard = new_context;
```

Comment:
We already got this config mutable, why we need to recreate the whole session_context here?